Add directives to automatically allocate free space

unknownbrackets opened the pull request:
This adds a few directives:

 * `.region`, like `.area` but set as shared.  Also always skips or fills to the end.
 * `.definesregion A,B,C` as a shorthand for `.org A :: .sharedarea B,C :: .endsharedarea`.
 * `.autoregion` which automatically allocates within regions.

Basic usage is explained in the Readme.  The basic concept here is, you may have pockets of free space (perhaps because you've deleted old data, removed functions, or shortened existing functions.)  You may then need to define a function.  Rather than searching for free space manually, and doing the math to see if it'll fit, this allows armips to take care of that for you.

`.autoregion` can take parameters, which may be necessary if it must be in a certain range (i.e. a Thumb bl range, etc.)

-[Unknown]

unknownbrackets commented on Readme.md:
````diff
@@ -371,16 +371,34 @@ Optionally, a second parameter can be given. The remaining free size of the area
+By default, `.autoarea` will allocate to any shared area with sufficient space.  It can be limited to a specific range if necessary:
 
 ```
 .autoarea @TextStart,@TextStart+@TextEnd
````
Oops, typo.  I had meant `@TextLen`.  Thanks for catching.

-[Unknown]

unknownbrackets commented on Readme.md:
````diff
@@ -371,16 +371,34 @@ Optionally, a second parameter can be given. The remaining free size of the area
 @TheAnswer:
   .byte 42
 .endautoarea
 ```
````
You're talking about this scenario, specifically, right?

```
.definesharedarea 0x082345678,0x1000,0x00

.autoarea
.func MyFunc
bx 14
.endfunc
.endautoarea

; Keep the party going.
dw 0x1337C0DE
```

In this scenario, it would keep writing after the end of the automatically allocated content, and there would be no error.  Specifically, it would write 0x1337C0DE to 0x08234567A, which might even be overwritten by another later `.autoarea`.

In theory it could move you to the end of that sharedarea (this is probably never useful) or perhaps restore the old position (possibly useful?)

Not sure if there's a great mechanism to mark the current address "invalid" and report as an error if you don't use org before assembly more.  That's probably the best idea, and least surprising.

The main purpose of `.area` in general is to keep yourself constrained, so I'd argue that this is a general problem.  It might be more obvious where you'll be after `.endarea`, but it's just as likely "wrong" to keep writing then as well.  Might see about adding this as a general error for all `.area` variants as a separate pull request.

-[Unknown]

Prof9 replied:
Writing more code after an `.area` is not an error IMO, because `.area` has a fixed size, therefore you should be at a predictable address after it ends.

IIRC this is how it works currently:
```
.org 0x2123456
.area 0x30
    // a bunch of code
    b       @end
.endarea
@end: // 0x2123486
    pop     r15
```

Blade2187 replied:
Yep, exactly that scenario. Couple points there...

1. That isn't how `.area` currently works; it defines a limitation, rather than enforcing that all that space must be taken up. The address you end up at is thus not predictable without evaluating what is within the `.area`. Unless, of course, you specify a byte with which to fill the unused space. (Does it fill this space before, or after, any `.autoarea`s consume the unused space? Ought to be before, I would think.)
2. `.autoarea` is a completely different kind of operation than `.area` or `.sharedarea`. It relocates the contained code somewhere else, which IIRC isn't something that anything else in armips previously did.
3. It's probably "wrong" to assemble after an `.area` without a fill byte specified, but not especially surprising. It's perfectly valid to assemble after an `.area` with a fill byte specified. It's definitely surprising, and bad behavior, if code/data assembled after an `.autoarea` is overwritten by a later `.autoarea`, or if it doesn't respect the bounds of the `.sharedarea`.

Given this and @unknownbrackets 's commentary, the behavior should probably be either to restore the old position, or throw a warning/error.

unknownbrackets replied:
For a `.sharedarea`, the fill parameter is applied after any allocation.  Consider:

```
.definesharedarea 0x08234567,0x1000,0x00
```

Any allocations into this shared area will be made first at 0x08234567, decreasing the remaining space to fill.  The filled bytes will be after.

As far as the order innternally within armips, it actually goes like this:

1. Validation pass 1: Area direct content and fill are in the area (same as before this change.)  Allocations are not placed anywhere yet.
2. Validation pass 2: Area direct content + allocations are put in the shared areas, and fill amount decreases (staying at end.)
3. Validation pass N: If moving the content into the shared area caused other things to recalculate (labels, etc.) they are re-evaluated until they stop changing (as per normal armips design.)

I've changed the behavior to reset to the previous position, but I still think any warnings/errors should be a separate change.  Although there are valid use cases (such as structs with zero fill) for using `.area` with fill, I think any warnings without fill/etc. should be considered more generally because they are not only relevant to any allocated area.

As far as the name - I'm not really tied to `.autoarea`, although I think it should be clear that it fits into areas.  I think most people will want to use it the same way I want to use it:

1. I'm currently using `.area` to replace existing functions in the game with functions friendly to localization.
2. In virtually every case, my implementation is more efficient (the compiler made some really terrible choices), so even after adding new features, VWF, centering logic, etc. 33% of the bytes I've replaced is still free.
3. The game has some ugly hacks, like printing "   's friends can be seen here" and then drawing the main character's name on top.  I replaced this with a simple control code i.e. "See stats for [NAME1] and friends."
4. The control code needed a function and didn't fit right in the text rendering loop (I'd already added so much there), but I knew I had several kilobytes of free space nearby.  I had to manually calculate where was a good spot, and I'd rather have software do the busy work for me.
5. I also didn't like having to put labels in obscure places to "fit" code in there, when it didn't semantically belong to that area (in the code, who cares about the output as long as it's a reachable area.)

So in my mind, it's still very conceptually linked to areas.  To me, they are like whiteout for the game's original data.  If I was specifically trying to fill a section with 0x00, I'd use `.fill`, because that conceptually makes more sense.

If we rename it, `.sharedarea` should probably be renamed to be consistent.

-[Unknown]

sp1187 commented:
I would like the ability to invalidate or specify shared areas in some way in case you have multiple overlays in a single assembly output file and need to restrict the autoareas to a specific overlay.

unknownbrackets commented:
It takes a parameter and is specific to the current output file.

So for example, if you have 3 output files, `.autoarea`s for the second file will only be allocated within `sharedarea`s in the second file.

In cases where you need it to be in a certain region of memory, you can use the parameters to force that.  For example, maybe you need it to be allocated anywhere within branch range of some source instruction - you could specify that.  You'd get an error if no free space was available where you required it.

I did consider trying to name the shared areas, but I think that opens up a lot of complexity.  It invites "tagging" the shared areas (see my use case above - maybe your overlay has 8 gaps you've punched out with rewritten functions or deleted data), and that gets into some syntax that feels very foreign among the rest of armips' syntax.

Personally I think specifying a range is the simplest.  I assume people would use variables as appropriate.

-[Unknown]

Prof9 commented:
I would very much like to have some sort of tagging functionality. Personally, for a project I'm working on I have a ton of free space in the ARM9 binary and not a whole lot in the actual overlay files themselves. So with the current implementation, it would be difficult if not impossible to use my ARM9 free space from an overlay.

Another reason would be `bl`s in a GBA game. Usually the .text section comes early in the ROM and all of the data blobs (graphics, text, etc.) are at the end. If you have free space near the .text section, you can use `bl`; otherwise, if you use free space at the end of the ROM, the jump distance is too great and you need to use `bx` instead. With tagging, you could tag `sharedarea`s as .text and then when you use `autoarea`, you can specify that you want one of the `sharedarea`s tagged with .text.

I don't think tags are all that foreign from the rest of armips' syntax seeing as we also have stuff like `.definelabel`, which takes an undefined symbol as a parameter. You can do the same kind of thing with macros. But if it makes things easier, you could consider using numbers as tags instead, then you could simply `.definelabel` all the tag numbers you want to use in order to have symbolic names.

unknownbrackets commented:
In my case, I've defined [22 areas](https://github.com/unknownbrackets/tomatotrans/search?q=endarea&unscoped_q=endarea) which doesn't even seem like that many.  I think all, or at least 20, are in the text section.

I definitely would not want to tag each one.  I'd rather say:

```
TextSegStart equ 0x000209D8
TextSegEnd equ 0x000F9600
```

Yes, I'd have to specify this for every autoarea, but I'd have to do that either way.

The complexity comes in on multiple tags.  Is it `.sharedarea 0x000F9600-.,0x00,tag1,tag2,tag3`?  What happens if I'm not filling?  Is it `.tagarea tag1 :: .tagarea tag2 :: .tagarea tag3` then?  What's the real utility of this over specifying a valid range using equ defines like above?  By adding a lot more syntax, are things easier, or is there just more I have to know to use it?

The `bl` example is really a great counterexample for tagging.  Since it can be +/- 4MB from the caller, doing it based on tag is limiting - maybe I have some space that is 1 MB later in the ROM (let's say by moving other data), and I can `bl` it just fine.  I could just say `.autoarea 0,0x00400000` and be done.

A related example is `b`, which has a range of 2 KB.  This is actually much more relevant for me, so let's give a real example.

I have a function that draws text.  Originally, this function could rely on length of text = number of 8x8 tiles to clear (i.e. to erase old text.)  With the VWF, this isn't true - it could easily be 2x as many characters as tiles.  In my case, I have a few helpers that look like this:

```
; Forces clear to 8, which is common.
.func CopyString8x8Clear8
mov r0,8
; Fall through to CopyString8x8ClearR0.
.endfunc
; This allows quick specification of clear width.
.func CopyString8x8ClearR0
ldr r1,=MFontClearSize
; Shorts to clear.
lsl r0,r0,4
strh r0,[r1]
b CopyString8x8ToVRAM
.endfunc

; Clear width in r0, pixel width in r2.
.func CopyString8x8CenterR0
ldr r1,=MFontClearSize
; Multiply by 8 to get pixel clear width.
lsl r3,r0,3
sub r3,r3,r2
lsr r3,r3,1
; Okay, now put that in the x override.
strb r3,[r1,MFontXOffset-MFontClearSize]

ldr r3,=CopyString8x8ClearR0+1
bx r3
.pool
.endfunc
```

In this case, I ended up using a bx just as you described.  But I probably DO have space within 2KB of `CopyString8x8ClearR0` and `CopyString8x8ToVRAM`, I just didn't want to manually find it.

Rather than using an army of tags to solve that (and each `b` I want), I'd again much rather specify a range: `.autoarea CopyString8x8ToVRAM-2048,CopyString8x8ToVRAM+1536`.  If my assumption that I have sufficient nearby space is wrong, it'll tell me.

As for managing overlays, it's more a question of code organization.  If it's possible to define helpers/functions that can go in the asm files that populate the ARM9 code, you could still have those auto allocate within ARM9.  I'm not that familiar with your use case, so I'm not sure if there's a good reason to group the code into the same source files as writes to the overlays.  But I do think it would become confusing if a source file wrote to multiple output files in a non-obvious way.

-[Unknown]

Prof9 commented:
Ah, I didn't realize you can constrain `.autoarea` to a specific memory range. I guess that's probably fine for my use case. That means the starting/ending address of the autoarea is, effectively, a single tag. And I could `.close` the overlay, `.open` the ARM9 binary, write my autoarea, `.close` the ARM9 binary and (if necessary) re-`.open` my overlay again. Slightly cumbersome, but not too bad.

I do think tags could still have a benefit here, because it would allow you to create free space in both the overlay and ARM9 file, and use the overlay free space first (since it can only be accessed from that specific overlay) before falling back on the ARM9 free space (which is accessible from any overlay). Otherwise, you would still be managing your `.autoarea`s manually to some degree.

Perhaps this could be realized by allowing the user to specify multiple ranges for `.autoarea`, and choose the first one that fits? Although this doesn't really solve the problem of the areas being in two separate physical files...

unknownbrackets commented:
Well, ideally, I'd like my `b` example above to be automatic, where it would just place it somewhere the content would validate.  But I'm also trying to avoid making perfect the enemy of improvement.  One change at a time.

I'm not sure if the same is realistic for overlays, since I assume armips has missing information about whether output file A can be known to be loaded while output file B is (my understanding is that these are, essentially, statically linked objects, but dynamically reloadable like `.so`s/DLLs.)  Even if it's obvious to you what is the always-loaded code, I assume armips (currently) has no idea.

Anyway, I'm not sure if anyone has a proposal for a better name than `autoarea`?

-[Unknown]

Blade2187 commented:
Perhaps slightly off-topic, but this particular use case seems risky to me... `b` should be used when you know for sure your hops are short-range. If you are using `b` to perform hops that might fail because you are relying on automatically-placed functions that reside elsewhere to be in a particular narrow range, you should probably just use `bx`. The ARM documentation does discuss using a _veneer_ to make an out-of-reach `bl` target accessible using `bx`, and avoiding having to do this is the sort of use case that I assumed was the intent, but I don't think that extends to `b`. Being able to automatically handle this would probably push armips towards compiler territory (?), which it is not.

Re: overflowing overlay additions into arm9, I'm not sure that's supported by armips, at least not automatically. I believe there is always only one output file at a time. Chances are that for now, you'd want to manually specify that certain code should go into the arm9, and then branch to it from the overlay. The opposite direction is of course right out, as multiple overlays share the same address space, and armips doesn't support knowing context of which one you mean short of actually opening it.

Re: tagging, I think it would probably solve several of these concerns that have to be manually specified for now, but can probably be pushed off to the future. For now probably best to just design the syntax to avoid impeding later implementation of tagging. Not sure on what that syntax should be, but what OP mentioned seemed reasonable.

Re: the name, I have no strong opinion on specific name, but I want to reiterate that `.autoarea` does an _entirely different type of thing_ than `.area`, `.sharedarea`, and `.definesharedarea`, and this is what motivated me to suggest a name change so as to not be similar to the others.

Prof9 commented:
`.autoarea` essentially just changes the memory address, right? So I'd expect the name to contain `org`, e.g. `.autoorg`.

I also think it would make sense if `.autoarea`, `.sharedarea` and `.definesharedarea` have a common element in their name. I'm not sure `.sharedorg` is a good name for `.autoarea`, though, since the memory address isn't what's being shared - it's the area. What about something like `.freeorg`, `.freearea` and `.definefreearea`?

Kingcom commented:
Currently it seems to use a First-Fit allocation algorithm. Is this good/appropriate enough or should we use a different one like Best-Fit?

unknownbrackets commented:
I think the allocation algorithm could be changed in a separate follow up change.  Best fit can be anything from simple to complex - could require more passes.  Simplest version of that would just be, pick the largest free area that matches requirements.

A more complex version (maybe still no extra passes, not sure) would be to analyze the available subareas/allocs/regions/whatever at the end of the first pass (in this pass, they are only sized.)  Main gain would be reducing wasted space if there's leftover.

I feel like reusing `org` is confusing because it has an end.

Perhaps we can just use a new word like `region`?  `.sharedregion` and `.autoregion`?  Then someone just has to know that `sharedregions` are, in some ways, similar to areas (in that they fill, and can have code in them which is not dynamically allocated - it's kept at the start.)  Perhaps even just `.region` and `.defineregion` could work.

> The ARM documentation does discuss using a veneer to make an out-of-reach

Well, if I'm hacking a GBA game to display a longer character name, or a PSP game to load a larger title image - I only care about following the ARM/MIPS published ABI guidelines to the extent that it's visible to other functions in the code (i.e. stack alignment.)  If I have limited bytes, I'm more than happy to use tail calls and `b` in this way.  Of course, if the code uses exceptions and might use a stack unwinder, I have to be slightly careful with that.

Luckily, if the `b` is not reachable, `armips` will fail to assemble, telling me I've made a mistake.  Exactly what I want.  I'm sure if it was a compiler (which it is indeed not), it might not allow my innocent violations of the ARM ABI.

-[Unknown]

Kingcom commented:
> 
> 
> I think the allocation algorithm could be changed in a separate follow up change. Best fit can be anything from simple to complex - could require more passes. Simplest version of that would just be, pick the largest free area that matches requirements.
I think that would actually be Worst-Fit. Each algorithm has benefits and drawbacks, though that approach can lead to being unable to allocate bigger chunks because all available areas are just barely too small. Using the area with the least available space is less susceptible to that. For now though, using the first available area should be sufficient. Until the next release we should be rather free in changing its implementation without worrying too much about compatibility.

> I feel like reusing `org` is confusing because it has an end.
I agree with this.

> Perhaps we can just use a new word like `region`? `.sharedregion` and `.autoregion`? Then someone just has to know that `sharedregions` are, in some ways, similar to areas (in that they fill, and can have code in them which is not dynamically allocated - it's kept at the start.) Perhaps even just `.region` and `.defineregion` could work.
I like this idea. It's different enough from area to prevent confusion, but it still makes sense that it conceptually does something similar.

unknownbrackets commented:
Well, it'd be worst fit depending on the scenario really, assuming they are allocated unordered.  As mentioned, it'd need to know the sizes of everything (to do an ordered allocation) to really do it right.

Renamed to regions (rebased to make it easier to see the code, also needed rebasing from recent changes anyway.)

Also added some tests.

-[Unknown]
